src/cmd/ksh93/bltins/hist.c

```diff
@@ -58,6 +58,13 @@ int	b_hist(int argc,char *argv[], Shbltin_t *context)
 	NOT_USED(argc);
 	if(!sh_histinit((void*)shp))
 		errormsg(SH_DICT,ERROR_system(1),e_histopen);
+
+	/* 'history' and 'r' builtins */
+	if(argv[0][4] == 'o')		/* hist<o>ry */
```

Comment:
This function may be called with an `argv[0]` that is shorter than 4 characters (`fc`, `r`), so this check is unsafe and causes unpredictable behaviour.

I would just change this line to:
```c
	if(!strcmp(argv[0],"history"))
```
After all, it's not as if we need to optimise performance here.

The check for `r` should be fine though, as it checks the first character only and this function is never called by another command that starts with `r`.

Reply:
I have applied the change that avoids `strcmp`.